tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2020 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Drivers for streaming reductions framework."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+from tensorflow_probability.python.experimental.mcmc import with_reductions
+from tensorflow_probability.python.experimental.mcmc import SampleDiscardingKernel
+from tensorflow_probability.python.experimental.mcmc import step_kernel
+from tensorflow.python.util import nest
+
+
+__all__ = [
+    'sample_fold',
+]
+
+
+def sample_fold(
+    num_steps,
+    current_state,
+    previous_kernel_results=None,
+    kernel=None,
+    reducers=None,
+    num_burnin_steps=0,
+    num_steps_between_results=0,
+    parallel_iterations=10,
+    seed=None,
+    name=None,
+):
+  """Finalizes `num_steps` Reducer steps with `kernel` samples.
+
+  This driver will appropriately apply `WithReductions` and
+  `SampleDiscardingKernel` over the given `kernel` and `reducers`. It will
+  then update reducer states `num_steps` times (perform `num_steps` calls to
+  the `Reducer`'s `one_step` method). An arbitrary collection of `reducers`
+  can be provided, and the resulting finalized statistic(s) will be returned
+  in an identical structure.
+
+  Args:
+    num_steps: Integer number of `Reducer` steps.
+    current_state: `Tensor` or Python `list` of `Tensor`s representing the
+      current state(s) of the Markov chain(s).
+    previous_kernel_results: A `Tensor` or a nested collection of `Tensor`s.
+      Warm-start for the auxiliary state needed by the given `kernel`.
+      If not supplied, `sample_fold` will cold-start with
+      `kernel.bootstrap_results`.
+    kernel: An instance of `tfp.mcmc.TransitionKernel` which implements one step
+      of the Markov chain.
+    reducers: A (possibly nested) structure of `Reducer`s to be evaluated
+      on the `kernel`'s samples. If no reducers are given (`reducers=None`), an
+      empty list will be returned in place of streaming calculations.
```

Comment:
It would be more stylish to return `None` rather than an empty list, since that more closely mirrors the input structure.